lib/ui/semantics.dart

```diff
@@ -641,6 +641,164 @@ class SemanticsFlag {
   }
 }
 
+/// A string that carries a list of [StringAttribute]s.
+class AttributedString extends NativeFieldWrapperClass2 {
+  /// Creates a attributed string.
+  ///
+  /// The [TextRange] in the [attributes] must be inside the length of the
+  /// [string].
+  AttributedString(
+    this.string, {
+    this.attributes,
+  }) : assert(() {
+         if (attributes != null) {
+           for (final StringAttribute attribute in attributes) {
+             assert(
+               string.length >= attribute.range.start &&
+               string.length >= attribute.range.end,
+               'The range in $attribute is outside of the string $string',
+             );
+           }
+         }
+         return true;
+       }()) {
+    _initAttributedString(this, string, attributes);
+  }
+
+  /// The plain string stored in the attributed string.
+  final String string;
+
+  // The attributes this string carries.
+  //
+  // The list must not be modified after this string is created.
+  final List<StringAttribute>? attributes;
+
+  void _initAttributedString(
+    AttributedString instance,
+    String string,
+    List<StringAttribute>? attributes,
+  ) native 'AttributedString_initAttributedString';
+
+  @override
+  AttributedString operator +(AttributedString other) {
+    if (string.isEmpty) {
+      return other;
+    }
+    if (other.string.isEmpty) {
+      return this;
+    }
+
+    // None of the strings is empty.
+    String newString = string + other.string;
+    List<StringAttribute>? newAttributes = attributes != null ? List<StringAttribute>.from(attributes!) : null;
+    if (other.attributes != null && other.attributes!.isNotEmpty) {
+      int offset = string.length;
+      newAttributes = newAttributes ?? <StringAttribute>[];
+      for (final StringAttribute attribute in other.attributes!) {
+        final TextRange newRange = TextRange(
+          start: attribute.range.start + offset,
+          end: attribute.range.end + offset,
+        );
+        final StringAttribute offsetedAttribute = attribute._copy(range: newRange);
+        newAttributes.add(offsetedAttribute);
+      }
+    }
+    return AttributedString(newString, attributes: newAttributes);
+  }
+
+  @override
+  String toString() {
+    return '$string${attributes != null && attributes!.isNotEmpty ? '\n(attributes: $attributes)': ''}';
```

Comment:
See https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#override-tostring for how this should be formatted.

ObjectRuntimeType isn't available here, I'd just directly use 'AttributedString'.